Source/Dafny/Resolver.cs

```diff
@@ -9625,295 +9688,888 @@ public class ResolveTypeReturn
+  if (branch is RBranchStmt) {
+    return new CStmt(new BlockStmt(tok, tok, ((RBranchStmt)branch).Body));
+  } else if (branch is RBranchExpr) {
+    return new CExpr(((RBranchExpr)branch).Body);
+  } else {
+     throw new System.ArgumentException("unexpected subclass of RBranch");
+  }
+}
+
+List<Statement> UnboxStmtContainer(SyntaxContainer con) {
+  if (con is CStmt) {
+    var r = new List<Statement>();
+    r.Add(((CStmt)con).Body);
+    return r;
+  } else {
+    throw new System.ArgumentException("expected a StmtContainer");
+  }
+}
+
+
+
+// let-bind a variable of name "name" and type "type" as "expr" on the body of "branch"
+void LetBind(RBranch branch, String name, Type type, bool isGhost, Expression expr, bool substitute = false) {
+    BoundVar oldVar = null;
+
+
+
+    // OS TODO: warning: substitution is currently non capture-avoiding
+    if(substitute){
+      if(expr is IdentifierExpr){
+        var idExpr = (IdentifierExpr)expr;
+        oldVar = new BoundVar(expr.tok, idExpr.Name, idExpr.Type);
+      } else if (expr is NameSegment) {
+        var nsExpr = (NameSegment)expr;
+        oldVar = new BoundVar(expr.tok, nsExpr.Name, nsExpr.Type);
+      } else {
+        substitute = false;
       }
+    }
 
-      var ctorsList = new List<Dictionary<string, DatatypeCtor>>();
-      if (s.Source.Type.AsDatatype is TupleTypeDecl) {
-        var udt = s.Source.Type.NormalizeExpand() as UserDefinedType;
-        foreach (Type typeArg in udt.TypeArgs) {
-          var t = PartiallyResolveTypeForMemberSelection(s.Tok, typeArg).NormalizeExpand() as UserDefinedType;
-          if (t != null && t.ResolvedClass is DatatypeDecl) {
-            dtd = (DatatypeDecl)t.ResolvedClass;
-            ctorsList.Add(datatypeCtors[dtd]);
-          } else {
-            ctorsList.Add(new Dictionary<string, DatatypeCtor>());
+    if (branch is RBranchStmt) {
+      if(substitute){
+        MatchCaseExprSubstituteCloner cloner = new MatchCaseExprSubstituteCloner(new BoundVar(expr.tok, name, type), oldVar);
+        List<Statement> list = new List<Statement>();
+
+        foreach (Statement ss in ((RBranchStmt)branch).Body) {
+          Statement clone = cloner.CloneStmt(ss);
+          list.Add(clone);
           }
-        }
+        ((RBranchStmt)branch).Body = list;
+      } else {
+        var cLVar = new LocalVariable(expr.tok, expr.tok, name, type, isGhost);
+        var cPat = new CasePattern<LocalVariable>(cLVar.EndTok, cLVar);
+        var cLet = new LetStmt(cLVar.Tok, cLVar.Tok, cPat, expr);
+        ((RBranchStmt)branch).Body.Insert(0, cLet);
+      }
+    } else if (branch is RBranchExpr) {
+      if(substitute){
+        MatchCaseExprSubstituteCloner cloner = new MatchCaseExprSubstituteCloner(new BoundVar(expr.tok, name, type), oldVar);
+        ((RBranchExpr)branch).Body = cloner.CloneExpr(((RBranchExpr)branch).Body);
+      } else {
+      var cBVar = new BoundVar(expr.tok, name, type);
```

Comment:
Please check all formatting before doing the PR. I will stop commenting on code style in this file.